astropy/io/misc/pyarrow/csv.py

```diff
@@ -0,0 +1,377 @@
+        )
+    import pyarrow as pa
+    from pyarrow import csv
+
+    return pa, csv
+
+
+def convert_pa_string_array_to_numpy(str_arr: "pa.ChunkedArray") -> np.ndarray:
+    """
+    Convert a PyArrow ChunkedArray of strings to a NumPy array.
+
+    Parameters
+    ----------
+    str_arr : pa.ChunkedArray
+        A PyArrow ChunkedArray containing string data.
+
+    Returns
+    -------
+    np.ndarray
+        A NumPy array containing the string data from the input ChunkedArray.
+        If the input ChunkedArray contains null values, the returned array will
+        be a masked array with nulls masked out.
+    """
+    # Check if string_array has any nulls
+    has_null = str_arr.null_count > 0
+
+    # Replace nulls with an empty string
+    str_arr_filled = str_arr.fill_null("") if has_null else str_arr
+
+    # Convert to NumPy array with fixed-length Unicode dtype
+    np_array = str_arr_filled.to_numpy().astype(str)
+
+    if has_null:
+        mask = str_arr.is_null().to_numpy()
+        np_array = np.ma.array(np_array, mask=mask, copy=False)
+
+    return np_array
+
+
+def convert_pa_array_to_numpy(arr):
+    """
+    Convert a PyArrow array to a NumPy array.
+
+    Parameters
+    ----------
+    arr : pyarrow.Array
+        The PyArrow array to be converted.
+
+    Returns
+    -------
+    np.ndarray or np.ma.MaskedArray
+        The converted NumPy array. If the input array contains null values,
+        a masked array is returned with nulls masked out.
+
+    Notes
+    -----
+    - If the input array is of string type, it delegates the conversion to
+      `convert_pa_string_array_to_numpy`.
+    - If the input array contains null values, they are replaced with 0 before
+      conversion, and a masked array is returned with the null positions masked.
+    """
+    pa, _ = get_pyarrow_csv()
+
+    if pa.types.is_string(arr.type):
+        return convert_pa_string_array_to_numpy(arr)
+
+    # Check if array has any nulls
+    has_null = arr.null_count > 0
+
+    # Replace nulls with an empty string
+    arr_filled = arr.fill_null(0) if has_null else arr
+
+    # Convert to NumPy array with fixed-length Unicode dtype
+    np_array = arr_filled.to_numpy()
+
+    if has_null:
+        mask = arr.is_null().to_numpy()
+        np_array = np.ma.array(np_array, mask=mask, copy=False)
+    return np_array
+
+
+def convert_pa_table_to_astropy_table(table_pa: "pa.Table") -> "Table":
+    """
+    Convert a PyArrow Table to an Astropy Table.
+
+    Parameters
+    ----------
+    table_pa : pa.Table
+        The PyArrow Table to be converted.
+
+    Returns
+    -------
+    Table
+        Converted astropy Table.
+    """
+    from astropy.table import Table
+
+    columns = {}
+    for name, col in zip(table_pa.column_names, table_pa.itercolumns()):
+        col_np = convert_pa_array_to_numpy(col)
+        columns[name] = col_np
+    out = Table(columns, copy=False)
+    return out
+
+
+def strip_comment_lines(
```

Comment:
I guess a question will be whether this relatively slow copy is actually worth it, or whether it is better to just fall back to the default reader for this case. After all, the most common case, of comments at the start of the file, can already be dealt with. My sense would be not to do this in the initial PR.

Note in this respect that for things of interest to reading with `pyarrow`, it is not impossible that the file size is of order the amount of memory one has!

Reply:
For the moment let's leave this in. One idea I've had is writing a streaming FileLike class that strips comments on a line-by-line basis. My first attempt at this failed within pyarrow somehow and I didn't track it down, but this *should* be possible. Then you might take a hit on speed but not memory.

Also don't forget a potential use case of reading thousands of big (but not huge) CSV files. E.g. reading the raw GAIA archive which they make available in ECSV... though of course I'd like to write a `pyarrow.ecsv` reader!

Reply:
I did some simple profiling, creating a CSV file that is 100000 lines with a few comments at the top. Even with this current implementation, the `pyarrow.csv` reader is about 35% faster than the fast ASCII reader. 

There might also be ways to optimize comment handling, for instance doing a first pass looking for commented lines without keeping the entire file in memory. If the commented lines are all at the top (presumably the most common case) then the pyarrow reader can be set to just skip those lines.